paths/paths_windows.go

```diff
@@ -0,0 +1,140 @@
+	userInfo, err := getTokenUserInfo(token)
+	if err != nil {
+		return nil, err
+	}
+
+	primaryGroup, err := getTokenPrimaryGroupInfo(token)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UserSids{userInfo.User.Sid, primaryGroup.PrimaryGroup}, nil
+}
+
+// SetStateDirPerms applies a restrictive ACL to the directory specified by dirPath.
+// It sets the following security attributes on the directory:
+// Owner: The user for the current process;
+// Primary Group: The primary group for the current process;
+// DACL: Full control to the current user and to the Administrators group.
+//       (We include Administrators so that admin users may still access logs;
+//        granting access exclusively to LocalSystem would require admins to use
+//        special tools to access the Log directory)
+// Inheritance: The directory does not inherit the ACL from its parent.
+//              However, any directories and/or files created within this
+//              directory *do* inherit the ACL that we are setting.
+func SetStateDirPerms(dirPath string) error {
+	info, err := os.Stat(dirPath)
+	if err != nil {
+		return err
+	}
+	if !info.IsDir() {
+		return os.ErrInvalid
+	}
+
+	// We need the info for our current user as SIDs
+	sids, err := getCurrentUserSids()
+	if err != nil {
+		return err
+	}
+
+	// We also need the SID for the Administrators group so that admins may
+	// easily access logs.
+	administratorsGroupSid, err := windows.CreateWellKnownSid(windows.WinBuiltinAdministratorsSid)
+	if err != nil {
+		return err
+	}
+
+	// Munge the SIDs into the format required by EXPLICIT_ACCESS.
+	userTrustee := windows.TRUSTEE{nil, windows.NO_MULTIPLE_TRUSTEE,
+		windows.TRUSTEE_IS_SID, windows.TRUSTEE_IS_USER,
+		windows.TrusteeValueFromSID(sids.User)}
+
+	adminTrustee := windows.TRUSTEE{nil, windows.NO_MULTIPLE_TRUSTEE,
+		windows.TRUSTEE_IS_SID, windows.TRUSTEE_IS_WELL_KNOWN_GROUP,
+		windows.TrusteeValueFromSID(administratorsGroupSid)}
+
+	// We declare our access rights via this array of EXPLICIT_ACCESS structures.
+	// We set full access to our user and to Administrators.
+	// We configure the DACL such that any Files or directories created within
+	// dirPath will also inherit this DACL.
+	explicitAccess := []windows.EXPLICIT_ACCESS{
+		windows.EXPLICIT_ACCESS{
+			windows.GENERIC_ALL,
+			windows.SET_ACCESS,
+			windows.SUB_CONTAINERS_AND_OBJECTS_INHERIT,
+			userTrustee,
+		},
+		windows.EXPLICIT_ACCESS{
+			windows.GENERIC_ALL,
+			windows.SET_ACCESS,
+			windows.SUB_CONTAINERS_AND_OBJECTS_INHERIT,
+			adminTrustee,
+		},
+	}
+
+	dacl, err := windows.ACLFromEntries(explicitAccess, nil)
+	if err != nil {
+		return err
+	}
+
+	// We now reset the file's owner, primary group, and DACL.
+	// We also must pass PROTECTED_DACL_SECURITY_INFORMATION so that our new ACL
+	// does not inherit any ACL entries from the parent directory.
+	const setFlags windows.SECURITY_INFORMATION = windows.OWNER_SECURITY_INFORMATION |
```

Comment:
the type here is unnecessary. And given the small scope this name lives (one line), I'd just name it "flags", so:

```
   const flags = windows.OWNER_SECURITY_INFORMATION | ....
```

Reply:
Acknowledged.

Reply:
Actually this is giving me problems. If I do:
```
flags := windows.OWNER_SECURITY_INFORMATON | ...
```
Then the compiler complains about `flags`'s type when I pass it into `SetNamedSecurityInfo`.
Wrapping that with a conversion:
```
windows.SetNamedSecurityInfo(..., windows.SECURITY_INFORMATION(flags), ...)
```
Makes the compiler complain that `constant 2147483655 overflows int`

Unfortunately from what I can tell, this is because the constants in the `windows` package are not explicitly typed as `windows.SECURITY_INFORMATION` or `uint32`. `windows.PROTECTED_DACL_SECURITY_INFORMATION` is `0x80000000`.

Thoughts @bradfitz?

Reply:
> If I do:
> flags := windows.OWNER_SECURITY_INFORMATON | ...

I said `const flags = ` ... :) 

See https://go.dev/blog/constants ... they're one of the coolest parts of Go.

Reply:
Ah, there we go. Thanks.